### PR TITLE
Add envelope() method to Rectangle

### PR DESCRIPTION
**Pull request by BroderickCarlin**
Thank you for helping out with embedded-graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This PR adds a `envelope()` method to the core `Rectangle` primitive. This method is intended to be the counter operation to the existing `intersection()`. This was added to aid development of higher level graphics libraries that may wish to group primitives and convey information about items such as bounding boxes without having to explicitly implement this logic. 

I am also happy to take any suggestions for alternatives to the term `envelope` if individuals feel strongly.

### CODE REVIEW

File: core/src/primitives/rectangle/mod.rs [hunk 1]
```diff
@@ -246,6 +246,65 @@ impl Rectangle {
...
+    ///     "############",
+    /// ]);
+    /// # Ok::<(), core::convert::Infallible>(())
+    /// ```
+    pub fn encompass(&self, other: &Rectangle) -> Rectangle {
+        let top_left = self.top_left.component_min(other.top_left);
+        let bottom_right = self
+            .bottom_right()
+            .unwrap_or(self.top_left)
+            .component_max(other.bottom_right().unwrap_or(other.top_left));
```
Comment (rfuest) [inline core/src/primitives/rectangle/mod.rs:303]:
```suggestion
        let bottom_right = self.anchor_point(AnchorPoint::BottomRight)
            .component_max(other.anchor_point(AnchorPoint::BottomRight));
```

**Review by rfuest — CHANGES_REQUESTED** [review]
That's a useful addition, I've missed this feature before.

One thing that needs to be considered is how zero sized rectangles should be handled. At the moment the code handles a zero sized rectangle the same same way as a rectangle with size 1x1 pixels. I'm unsure if this is the best way to handle this or if it is better to ignore zero sized rectangles in the calculation. I think we should just pick one and document how the method behaves in that cases.

This [Wikipedia page](https://en.wikipedia.org/wiki/Minimum_bounding_rectangle) lists "minimum bounding rectangle", "bounding box" and "envelope" as common names for such a rectangle. IMO `envelope` or `enveloping` would be good names for the function, because "bounding box" is already used elsewhere in e-g.

File: core/src/primitives/rectangle/mod.rs (hunk 1 above)
Comment (BroderickCarlin) [reply core/src/primitives/rectangle/mod.rs:303]: Good example of why it's probably best if e-g provides such a method directly 😄

**Comment by BroderickCarlin** [comment]
went with `envelope()` to match `intersect()` such that both lacked the `-ing` suffix.

**Comment by BroderickCarlin** [comment]
In terms of zero-sized rectangles - my current thought is such that these items do have a location that they exist at - even if we can't represent their size. Treating them as 1x1 rectangles, as is currently done, appears the most intuitive to myself but I also confess there may be a use-case that I simply am overlooking. Happy to add the documentation either way

File: core/src/primitives/rectangle/mod.rs [hunk 2]
```diff
@@ -246,6 +246,64 @@ impl Rectangle {
         Rectangle::zero()
     }
 
+    /// Returns the minimum sized `Rectangle` that envelopes both `self` and `other`.
+    ///
+    /// # Examples
+    ///
+    /// ## Envelope
+    ///
```
Comment (rfuest) [inline core/src/primitives/rectangle/mod.rs:254]: The heading isn't needed, because there is only one example.
```suggestion
    /// ## Envelope
    ///
```

**Review by rfuest — CHANGES_REQUESTED** [review]

**Comment by rfuest** [comment]
> In terms of zero-sized rectangles - my current thought is such that these items do have a location that they exist at - even if we can't represent their size. Treating them as 1x1 rectangles, as is currently done, appears the most intuitive to myself but I also confess there may be a use-case that I simply am overlooking. Happy to add the documentation either way

OK, let's go this way. I think we should add another test for this case, to not only check that rectangles like `Rectangle::new(tl, Size::zero())`, but also something like `Rectangle::new(tl, Size::new(10, 0))` works as expected. And the documentation should state that if the rectangle size is 0 in any dimension it is treated as 1 by the `envelope` method.

**Comment by BroderickCarlin** [comment]
Went ahead and added a few more unit tests exercising zero sized rects and their interaction with this method

**Review by rfuest — CHANGES_REQUESTED** [review]
Some final nits.

File: core/src/primitives/rectangle/mod.rs [hunk 3]
```diff
@@ -246,6 +246,67 @@ impl Rectangle {
         Rectangle::zero()
     }
 
+    /// Returns the minimum sized `Rectangle` that envelopes both `self` and `other`.
+    ///
+    /// For the purpose of this method, zero-sized [`Rectangle`]s will be treated
+    /// equivelant to 1x1 [`Rectangle`]s. This is done so that the [`Rectangle`] is
```
[the remark this answers is not in the document]
Comment (BroderickCarlin) [reply core/src/primitives/rectangle/mod.rs:252]: It seems I need to confirm my spellchecker is enabled 😅

File: core/CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ### Changed
 
 - (technically breaking) [#731](https://github.com/embedded-graphics/embedded-graphics/pull/731) Bump MSRV to 1.71.1
+- [#733](https://github.com/embedded-graphics/embedded-graphics/pull/733) Add envelope() method to `Rectangle`
```
[the remark this answers is not in the document]
Comment (BroderickCarlin) [reply core/CHANGELOG.md:12]: Also moved this under an `Added` section

File: core/src/primitives/rectangle/mod.rs [hunk 5]
```diff
@@ -246,6 +246,67 @@ impl Rectangle {
         Rectangle::zero()
     }
 
+    /// Returns the minimum sized `Rectangle` that envelopes both `self` and `other`.
+    ///
+    /// For the purpose of this method, zero-sized [`Rectangle`]s will be treated
```
[the remark this answers is not in the document]
Comment (BroderickCarlin) [reply core/src/primitives/rectangle/mod.rs]: Tweaked the text here to better convey that a rectangle will be treated as having a minimum possible size of `1` along each specific axis, such that 0x0 is treated as 1x1 and 5x0 is treated as 5x1

File: core/CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfuest) [reply core/CHANGELOG.md:14]: Good catch.

**Review by rfuest — APPROVED** [review]
LGTM
